src/Lecture2.hs

```diff
@@ -414,4 +446,18 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding expr = varsToExpr (collectVars expr (0, []))
+  where
+    varsToExpr :: (Int, [String]) -> Expr
+    varsToExpr (0, x:xs) = foldr (Add . Var) (Var x) xs
+    varsToExpr (i, []) = Lit i
+    varsToExpr (i, xs) = Add (varsToExpr (i, [])) (varsToExpr (0, xs))
+
```

Comment:
Wow, that's a very cute and clever solution! 🤯 
I might borrow it for the recommended solutions to this lecture 😅 

I personally would implement a helper function to convert a list of variable names to not rely on patterns that can potentially change in the future. For example, when you call `varsToExpr (0, xs)` recursively, it wasn't immediately obvious to me that `xs` is not empty here 🤔 
```suggestion
    varsToExpr (0, x : xs) = addVars x xs
    varsToExpr (i, []) = Lit i
    varsToExpr (i, x : xs) = Add (Lit i) (addVars x xs)
    
    addVars :: String -> [String] -> Expr
    addVars x = foldr (Add . Var) (Var x)
```